src/databricks/sql/auth/oauth.py

```diff
@@ -0,0 +1,282 @@
+
+import requests
+from requests.exceptions import RequestException
+
+
+try:
+    from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
+except ImportError:
+    from http.server import BaseHTTPRequestHandler, HTTPServer
+
+logger = logging.getLogger(__name__)
+
+
+# This could use 'import secrets' in Python 3
+def token_urlsafe(nbytes=32):
+    tok = os.urandom(nbytes)
+    return base64.urlsafe_b64encode(tok).rstrip(b"=").decode("ascii")
+
+
+# This could be datetime.timezone.utc in Python 3
+class UTCTimeZone(tzinfo):
+    """UTC"""
+    def utcoffset(self, dt):
+        #pylint: disable=unused-argument
+        return timedelta(0)
+
+    def tzname(self, dt):
+        #pylint: disable=unused-argument
+        return "UTC"
+
+    def dst(self, dt):
+        #pylint: disable=unused-argument
+        return timedelta(0)
+
+
+# Some constant values
+OIDC_REDIRECTOR_PATH = "oidc"
+REDIRECT_PORT = 8020
+UTC = UTCTimeZone()
+
+
+def get_redirect_url(port=REDIRECT_PORT):
+    return f"http://localhost:{port}"
+
+
+def fetch_well_known_config(idp_url):
+    known_config_url = f"{idp_url}/.well-known/oauth-authorization-server"
+    try:
+        response = requests.get(url=known_config_url)
+    except RequestException as e:
+        logger.error(f"Unable to fetch OAuth configuration from {idp_url}.\n"
+                     "Verify it is a valid workspace URL and that OAuth is "
+                     "enabled on this account.")
+        raise e
+
+    if response.status_code != 200:
+        msg = (f"Received status {response.status_code} OAuth configuration from "
+               f"{idp_url}.\n Verify it is a valid workspace URL and "
+               "that OAuth is enabled on this account."
+               )
+        logger.error(msg)
+        raise RuntimeError(msg)
+    try:
+        return response.json()
+    except requests.exceptions.JSONDecodeError as e:
+        logger.error(f"Unable to decode OAuth configuration from {idp_url}.\n"
+                     "Verify it is a valid workspace URL and that OAuth is "
+                     "enabled on this account.")
+        raise e
+
+
+def get_idp_url(host):
+    maybe_scheme = "https://" if not host.startswith("https://") else ""
+    maybe_trailing_slash = "/" if not host.endswith("/") else ""
+    return f"{maybe_scheme}{host}{maybe_trailing_slash}{OIDC_REDIRECTOR_PATH}"
+
+
+def get_challenge(verifier_string=token_urlsafe(32)):
+    digest = hashlib.sha256(verifier_string.encode("UTF-8")).digest()
+    challenge_string = base64.urlsafe_b64encode(digest).decode("UTF-8").replace("=", "")
+    return verifier_string, challenge_string
+
+
+# This is a janky global that is used to store the path of the single request the HTTP server
+# will receive.
+global_request_path = None
+
+
+def set_request_path(path):
+    global global_request_path
+    global_request_path = path
+
+
+class SingleRequestHandler(BaseHTTPRequestHandler):
+    RESPONSE_BODY = """<html>
+<head>
```

Comment:
Maybe put this into an external file where it will be easier to view and edit? (Yes, I know this is what we did for databricks-cli, but that was kind of treated as experimental.)

Reply:
thanks for the comment addressed. also I got rid of the global variable.